--- a/make/hotspot/lib/JvmMapfile.gmk	2024-01-09 21:18:15.000000000 +0000
+++ b/make/hotspot/lib/JvmMapfile.gmk	2024-03-17 09:23:34.068511556 +0000
@@ -53,7 +53,7 @@
 # platform dependent.
 
 ifeq ($(call isTargetOs, linux), true)
-  DUMP_SYMBOLS_CMD := $(NM) --defined-only *.o
+  DUMP_SYMBOLS_CMD := $(NM) --defined-only --extern-only *.o
   ifneq ($(FILTER_SYMBOLS_PATTERN), )
     FILTER_SYMBOLS_PATTERN := $(FILTER_SYMBOLS_PATTERN)|
   endif
